### PR TITLE
Added eslint-plugin-tsdoc

### PR DESCRIPTION
**Pull request by tasneemkoushar**
**What kind of change does this PR introduce?**

It checks and tells us about the errors and problems associated with **tsdoc** documentation. It also fixes Graphql-eslint plugin configuration.

**Summary**

This update is necessary as when contributors start contributing to talawa-api, this check will let them know about the syntax and other errors in their documentation along with their code contribution.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by palisadoes** [comment]
1. Please update CONTRIBUTING.md explaining the use of this option.
1. Please also tie your PR to an issue

**Comment by xoldyckk** [comment]
@tasneemkoushar @palisadoes 

The errors are because of adding this to the config.
```
    "plugin:@typescript-eslint/recommended",
```

Seems like
```
"plugin:@typescript-eslint/eslint-recommended"
```
on its own wasn't catching many issues which are considered bad in typescript code.

For example use of 
```
@ts-ignore
```
to ignore typescript errors. But the places in code where this is used are either incorrectly implemented or making using packages whose owners do not provide type definitions.

**Comment by palisadoes** [comment]
@xoldyckk 

1. Should we create issues to remove the `@ts-ignore` statements? Now would be the time as we should be seeing many new contributors this week.
1. Do we need the `lib/` directory any more?
1. Do you have any other suggestions regarding this?

**Comment by xoldyckk** [comment]
> @xoldyckk
> 
> 1. Should we create issues to remove the `@ts-ignore` statements? Now would be the time as we should be seeing many new contributors this week.
> 2. Do we need the `lib/` directory any more?
> 3. Do you have any other suggestions regarding this?

`lib` directory doesn't exist anymore, I moved everything inside it to `src` directory, makes it much cleaner. `@ts-ignore` is just one of the issues, there are many others which are considered bad in typescript code. When I tried linting it with `@typescript-eslint/recommended` eslint plugin I had like more than a thousand lint errors.

**Comment by palisadoes** [comment]
> > @xoldyckk
> > 
> > 1. Should we create issues to remove the `@ts-ignore` statements? Now would be the time as we should be seeing many new contributors this week.
> > 2. Do we need the `lib/` directory any more?
> > 3. Do you have any other suggestions regarding this?
> 
> `lib` directory doesn't exist anymore, I moved everything inside it to `src` directory, makes it much cleaner. `@ts-ignore` is just one of the issues, there are many others which are considered bad in typescript code. When I tried linting it with `@typescript-eslint/recommended` eslint plugin I had like more than a thousand lint errors.

Now is the time to get GSoC candidates to help fix the errors. The organizations are announced on Wednesday afternoon UTC. We could create a small number of fixit bug issues to correct each category of error in anticipation of us rolling out this feature.

What would those categories be, and would you be able to create those issues by the middle of the week?

**Comment by xoldyckk** [comment]
> > > @xoldyckk
> > > 
> > > 1. Should we create issues to remove the `@ts-ignore` statements? Now would be the time as we should be seeing many new contributors this week.
> > > 2. Do we need the `lib/` directory any more?
> > > 3. Do you have any other suggestions regarding this?
> > 
> > 
> > `lib` directory doesn't exist anymore, I moved everything inside it to `src` directory, makes it much cleaner. `@ts-ignore` is just one of the issues, there are many others which are considered bad in typescript code. When I tried linting it with `@typescript-eslint/recommended` eslint plugin I had like more than a thousand lint errors.
> 
> Now is the time to get GSoC candidates to help fix the errors. The organizations are announced on Wednesday afternoon UTC. We could create a small number of fixit bug issues to correct each category of error in anticipation of us rolling out this feature.
> 
> What would those categories be, and would you be able to create those issues by the middle of the week?

Most issues are warnings not errors. 

Errors are less than 100 in number. Some of them can be fixed but not all. Because they're problems on the part of packages talawa-api uses.

Warnings are generated mostly because of use of `any` keyword. `any` keyword basically does what it says, makes the type of things `any`. This makes typescript the same as javascript(no concrete type). With `any` typescript cannot guarantee type safety.

**Comment by palisadoes** [comment]
@tasneemkoushar How do you plan to proceed?

**Comment by tasneemkoushar** [comment]
@

> @tasneemkoushar How do you plan to proceed?

I agree with @xoldyckk , we can create issues for GSOC candidates to help fix the errors, and ignore the ones that cannot be fixed for now.

**Comment by palisadoes** [comment]
Thanks, when could that begin?

**Comment by tasneemkoushar** [comment]
> Thanks, when could that begin?

@xoldyckk Are you likely to be free soon?

**Comment by xoldyckk** [comment]
@palisadoes Type errors cannot be fixed until deprecated packages are removed.

**Comment by noman2002** [comment]
@palisadoes @tasneemkoushar Should we close this ?

**Comment by palisadoes** [comment]
@tasneemkoushar Can this check be limited to just the formatting of comments?

**Comment by tasneemkoushar** [comment]
> @tasneemkoushar Can this check be limited to just the formatting of comments?

@noman2002  Do not close this. I just have loads of tasks stacked up. @palisadoes I need to figure out if we can limit it to comments only.

**Comment by tasneemkoushar** [comment]
Hi @palisadoes @xoldyckk 
I came up with a solution. I made a point of disabling all lint errors and warnings.
Now that we can target the specific type of issues we want to address rather than just fixing lint errors in the code, the process will go more quickly and easily.

Statistically, for this branch, we currently have:- 
56 params and tag warnings (TSDOC),
67 errors, 1434 warnings (Typescript Rules),
3 errors, 22 warnings (Extension Rules),
0 Warnings (Formatting Rules),
0 Warnings (Deprecated Rules)

We can create issues to fix these warnings and errors.
You can read more about the rules [here](https://typescript-eslint.io/rules/).

**Comment by xoldyckk** [comment]
@tasneemkoushar Did you just turn off all typescript-eslint rules there are?

File: .eslintrc.json [hunk 1]
```diff
@@ -6,6 +6,7 @@
   "extends": [
     "eslint:recommended",
     "plugin:@typescript-eslint/eslint-recommended",
```
Comment (xoldyckk) [inline .eslintrc.json:8]: Remove `plugin:@typescript-eslint/eslint-recommended`, it's not needed.

File: .eslintrc.json [hunk 2]
```diff
@@ -6,6 +6,7 @@
...
+    "plugin:@typescript-eslint/recommended",
     "plugin:prettier/recommended"
```
Comment (xoldyckk) [inline .eslintrc.json:10]: Remove `plugin:prettier/recommended`, add just `prettier` here. Prettier recommends not to use the plugin anymore.

File: .eslintrc.json [hunk 3]
```diff
@@ -21,10 +22,16 @@
...
   "parserOptions": {
+    "project": "./tsconfig.json",
+    "tsconfigRootDir": ".",
     "ecmaVersion": "latest",
     "sourceType": "module"
   },
-  "plugins": ["@typescript-eslint"],
+  "plugins": [
+    "@typescript-eslint",
+    "@typescript-eslint/eslint-plugin",
```
Comment (xoldyckk) [inline .eslintrc.json:32]: Remove `@typescript-eslint/eslint-plugin`

File: .eslintrc.json [hunk 4]
```diff
@@ -45,6 +52,151 @@
         "allowTemplateLiterals": true,
         "avoidEscape": true
```
Comment (xoldyckk) [inline .eslintrc.json:53]: Remove the rules `eqeqeq`, `linebreak-style`, `no-unused-vars` and `quotes`. Eslint and prettier come with sensible defaults already, it's just overhead to define our own.

File: .eslintrc.json [hunk 5]
```diff
@@ -45,6 +52,151 @@
...
+    "@typescript-eslint/space-before-function-paren": "off",
+    "@typescript-eslint/space-infix-ops": "off",
+    "@typescript-eslint/type-annotation-spacing": "off",
+
+    // Deprecates Rules
+    "@typescript-eslint/no-duplicate-imports": "off",
+    "@typescript-eslint/no-implicit-any-catch": "off",
+    "@typescript-eslint/no-parameter-properties": "off",
+    "@typescript-eslint/sort-type-union-intersection-members": "off",
+    "@typescript-eslint/camelcase": "off"
```
Comment (xoldyckk) [inline .eslintrc.json:200]: Did you just turn off all the typescript eslint rules there are in existence? Why would you do that? Shouldn't you only turn off rules which are causing the errors and warnings?

File: package.json [hunk 6]
```diff
@@ -99,6 +100,7 @@
     "eslint": "^8.26.0",
     "eslint-config-prettier": "^8.5.0",
     "eslint-plugin-prettier": "^4.2.1",
```
Comment (xoldyckk) [inline package.json:102]: Remove `eslint-plugin-prettier` not needed.

File: package.json [hunk 7]
```diff
@@ -13,6 +13,7 @@
     "typecheck": "graphql-codegen && tsc --noEmit --pretty",
     "lint:check": "eslint . --max-warnings=0",
     "lint:fix": "eslint . --fix",
+    "lint:tsdoc": "eslint -f unix \"src/**/*.{ts,tsx}\"",
```
Comment (xoldyckk) [inline package.json:16]: Why does this need a seperate script? The `lint:fix` and `lint:check` do the same thing.

**Comment by palisadoes** [comment]
Please fix the conflicting files

**Comment by tasneemkoushar** [comment]
> Please fix the conflicting files

@palisadoes  Fixed conflicting files!

**Comment by tasneemkoushar** [comment]
> Is this ready for review now?

Yes

**Comment by palisadoes** [comment]
@tasneemkoushar Please resolve the conversations. @xoldyckk Please review.

**Comment by xoldyckk** [comment]
@tasneemkoushar Graphql-eslint plugin isn't properly configured, fix it within this PR.

Replace the `overrides` list:-
```
  "overrides": [
    {
      "files": ["*.ts"],
      "processor": "@graphql-eslint/graphql"
    },
    {
      "files": ["*.graphql"],
      "parser": "@graphql-eslint/eslint-plugin",
      "plugins": ["@graphql-eslint"]
    },
    {
      "files": ["tests/**/*"],
      "rules": {
        "no-restricted-imports": "off"
      }
    }
  ],
```
with this
```
  "overrides": [
    {
      "files": ["./src/typeDefs/**/*.ts"],
      "processor": "@graphql-eslint/graphql"
    },
    {
      "files": ["./src/typeDefs/**/*.graphql"],
      "parser": "@graphql-eslint/eslint-plugin",
      "plugins": ["@graphql-eslint"]
    },
    {
      "files": ["tests/**/*"],
      "rules": {
        "no-restricted-imports": "off"
      }
    }
  ]
```

**Comment by palisadoes** [comment]
@xoldyckk Please take another look

**Comment by xoldyckk** [comment]
@tasneemkoushar After this PR is merged make the relevant issues for contributers to fix lint errors/warnings for which you have turned off those eslint rules. Write a thorough series of steps how they'll need to look for those errors/warnings and fix them.

**Review by xoldyckk — APPROVED** [review]

**Pull request merged**